### PR TITLE
HTTP: Refactor into modern C++

### PR DESCRIPTION
**Pull request by tobiaskohlbau**
I'm in the process of refactoring some network parts and would like to receive early comments.

I've started with replacing jansson by https://github.com/nlohmann/json and working on a more modern http implementation.

Ignore the changes within `openrct2-cli` they're only for testing and will be removed within final PR.

Please comment on your thoughts and concerns.

### CODE REVIEW

**Comment by Gymnasiast** [comment]
Could you explain why you replaced jansson?

**Comment by IntelOrca** [comment]
I wouldn't mix changing json library and refactoring network requests. Do them in isolation.

**Comment by tobiaskohlbau** [comment]
@Gymnasiast I've decided against jansson cause it's c style and an additional dependency which could be dropped by including a modern, single file only version for JSON.

@IntelOrca I've decided in this way as IMHO a http library shouldn't decode JSON it's task of the user. It's not mandatory to use it, but if I get positive feedback for dropping jansson I would change the handling in conjunction with touching the network stuff. I've to replace it anyway as the design of new http library is different.

**Comment by janisozaur** [comment]
I'm not fully sold on switching away from jansson, but I can see the benefit a good API can provide.

From my point of view:
Benefits:
* Cleaner API leading to fewer errors and gotchas\*
* One less library to worry about
  - Especially in context of building with vcpkg on AppVeyor, jansson can take long time to configure

Drawbacks:
* Increase in compilation time
* We already have a JSON library

\* I say this not having looked at the new library's API yet, but I have heard good things about it and usually "modern" libraries have better interfaces.

**Comment by tobiaskohlbau** [comment]
I've stripped out the json part and finished refactoring.

File: src/openrct2/network/Http.cpp [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#pragma region Copyright (c) 2014-2017 OpenRCT2 Developers
+#pragma region Copyright(c) 2014 - 2017 OpenRCT2 Developers
```
Comment (janisozaur) [inline src/openrct2/network/Http.cpp]: 2018 – same for other files.

**Comment by janisozaur** [comment]
@tobiaskohlbau there are failures on Travis mostly because you renamed `http.h` to `Http.h` but not updated the `#include`s and some functions are missing previous declaration – all functions have to have explicit declarations or be marked as static/moved into anonymous namespaces. Can you please check Travis and address those issues?

Pay attention there are jobs with network support disabled and with http support disabled too.

File: src/openrct2-ui/windows/ServerList.cpp [hunk 2]
```diff
@@ -683,29 +683,29 @@ static void fetch_servers()
...
+    request.header["Accept"] = "application/json";
     status_text = STR_SERVER_LIST_CONNECTING;
-    http_request_async(&request, fetch_servers_callback);
+    http::DoAsync(request, fetch_servers_callback);
 #endif
 }
 
 #ifndef DISABLE_HTTP
-static void fetch_servers_callback(http_response_t* response)
+static void fetch_servers_callback(http::Response response)
```
Comment (IntelOrca) [inline src/openrct2-ui/windows/ServerList.cpp]: This should pass response by reference (const reference if possible). Often I delete the copy constructor of a class to check I am not copying unnecessarily.

File: src/openrct2/network/Http.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    Client();
+    ~Client();
+
+    Response Do(Request req);
+
+private:
+    static std::atomic<int> refcount;
+};
+
+Response Do(Request req);
```
Comment (IntelOrca) [inline src/openrct2/network/Http.h]: `req` could be passed by const reference.

File: src/openrct2/network/Http.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+struct Client
+{
+    Client();
+    ~Client();
+
+    Response Do(Request req);
+
+private:
+    static std::atomic<int> refcount;
```
Comment (IntelOrca) [inline src/openrct2/network/Http.h]: Avoid `int` as it is not guaranteed to be the same size on all platforms.

Comment (janisozaur) [reply src/openrct2/network/Http.h]: Use our typedefed types instead:

https://github.com/OpenRCT2/OpenRCT2/blob/e602162fdb6ad2041d236ca10344ef3366bd895b/src/openrct2/common.h#L38-L45

**Comment by tobiaskohlbau** [comment]
@janisozaur thanks for pointing out. Just as you commented I've started checking if Linux builds are fine as well, of course they weren't. Hopefully now everything builds.

File: src/openrct2/Context.cpp [hunk 5]
```diff
@@ -1,4 +1,4 @@
-#pragma region Copyright (c) 2014-2017 OpenRCT2 Developers
```
Comment (Gymnasiast) [inline src/openrct2/Context.cpp]: Taking out this space is unnecessary. Just modify the end year.

Comment (tobiaskohlbau) [reply src/openrct2/Context.cpp]: See my edit ;-)

**Comment by tobiaskohlbau** [comment]
I've fixed all reviews, WIP state should be removed and if no further concerns arise this PR could be merged.

File: src/openrct2/network/Http.cpp [hunk 6]
```diff
@@ -14,313 +14,234 @@
...
-        {
-            json_decref(Body.Json);
-        }
-        else
-        {
-            delete Body.Buffer;
-        }
-    }
-};
+static size_t writeData(char * ptr, size_t size, size_t nmemb, void * userdata)
```
Comment (janisozaur) [inline src/openrct2/network/Http.cpp]: `ptr` should be `const` and while at it maybe call it `src`?

File: src/openrct2/network/Http.cpp [hunk 7]
```diff
@@ -14,313 +14,234 @@
...
 
-typedef struct read_buffer {
-    char *ptr;
-    size_t length;
-    size_t position;
-} read_buffer;
+    auto mem = std::unique_ptr<char[]>(new char[realsize + 1]);
+    memcpy(mem.get(), ptr, realsize);
+    mem.get()[realsize] = '\0';
+    res->body += std::string(mem.get());
```
Comment (janisozaur) [inline src/openrct2/network/Http.cpp]: is there a reason for this `unique_ptr`? Can't this simply be `std::string(ptr, ptr + realsize)`? It should handle the case where it is not null terminated as well: https://wandbox.org/permlink/SQW8RHtdLu9tYwpQ

**Comment by devnoname120** [comment]
Note that this code is not thread-safe because of a race-condition in the initialization of cURL.

```Cpp
Client::Client()
{
    if (++refcount == 1)
        curl_global_init(CURL_GLOBAL_DEFAULT);
}

Client::~Client()
{
    if (--refcount == 0)
        curl_global_cleanup();
}
```

Let's say take the following scenario:
- Initially `refcount = 1`
- `Client::Client()` is called by thread A.
- `Client::~Client()` is called by thread B.
- B: `if (--refcount == 0)` gets executed, `refcount == 0` now.
- A: `if (++refcount == 1)` gets executed, `refcount == 1` now.
- A: `curl_global_init(CURL_GLOBAL_DEFAULT);` gets executed even though it's already initialized.
- B: `curl_global_cleanup();` gets executed, so cURL is now initialized.

In this scenario `refcount == 1` and cURL is used by thread A even though cURL is uninitialized.

File: src/openrct2/network/Http.cpp (hunk 7 above)
Comment (tobiaskohlbau) [reply src/openrct2/network/Http.cpp]: Didn't know about that opportunity. C styled strings I don't know inside out. I very much appreciate every tips.

File: src/openrct2/network/Http.cpp (hunk 6 above)
Comment (tobiaskohlbau) [reply src/openrct2/network/Http.cpp]: I've copied that straight out of the examples, if I've confirmed that everything builds with const char within function signature I'm fine.

**Comment by tobiaskohlbau** [comment]
@devnoname120 thanks for pointing this out. I wasn't aiming towards thread safety as libcurl is neither, but you got me and I changed my mind. Newest change uses a HTTP object as lifetime handling.

**Comment by tobiaskohlbau** [comment]
I've rebased this PR. Let me know if there are pending changes which resolve in not merging this.

File: src/openrct2/network/Http.cpp [hunk 8]
```diff
@@ -14,313 +14,214 @@
...
-    if (request.Type == HTTP_DATA_JSON && request.Body.Json != nullptr) {
-        readBuffer.ptr = json_dumps(request.Body.Json, JSON_COMPACT);
-        readBuffer.length = strlen(readBuffer.ptr);
-        readBuffer.position = 0;
-    } else if (request.Type == HTTP_DATA_RAW && request.Body.Buffer != nullptr) {
-        readBuffer.ptr = request.Body.Buffer;
-        readBuffer.length = request.Size;
-        readBuffer.position = 0;
-    }
+    CURL *                curl = curl_easy_init();
```
Comment (janisozaur) [inline src/openrct2/network/Http.cpp]: do we really need this on each request? Sounds expensive.

Comment (janisozaur) [reply src/openrct2/network/Http.cpp]: ok, that looks like the intended usage and we did that earlier anyway.

File: src/openrct2/network/Http.cpp [hunk 9]
```diff
@@ -14,313 +14,214 @@
...
         *outData = nullptr;
         return 0;
     }
 
-    size_t dataSize = response->size - 1;
-    void * data = malloc(dataSize);
-    if (data == nullptr) {
-        dataSize = 0;
+    size_t dataSize = res.body.size() - 1;
+    void * data     = malloc(dataSize);
```
Comment (janisozaur) [inline src/openrct2/network/Http.cpp]: `Memory::Allocate`

File: src/openrct2/network/Http.h [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+ * A full copy of the GNU General Public License can be found in licence.txt
+ *****************************************************************************/
+#pragma endregion
+
+#ifndef OPENRCT2_NETWORK_HTTP_H_
+#define OPENRCT2_NETWORK_HTTP_H_
+
+#ifndef DISABLE_HTTP
+
+#include <atomic>
```
Comment (janisozaur) [inline src/openrct2/network/Http.h]: Doesn't look like this header is needed here?

File: src/openrct2/network/Http.cpp (hunk 9 above)
Comment (AaronVanGeffen) [reply src/openrct2/network/Http.cpp]: Don't use Memory::Allocate — we're moving away from it entirely, cf. #7066

Comment (janisozaur) [reply src/openrct2/network/Http.cpp]: Hmm, there's https://github.com/OpenRCT2/OpenRCT2/blob/9a362a8aa382334f04b88938490990d5adb00021/src/openrct2/core/MemoryStream.cpp#L59 which is not touched by the PR you mentioned.

@IntelOrca ?

**Comment by janisozaur** [comment]
Some [relevant discussion](https://gitter.im/OpenRCT2/OpenRCT2?at=5a6665c5e0141226507b41ad):
```
matrixbot @matrixbot
noname120 Can HTTP be instantiated multiple times?
noname120 https://github.com/OpenRCT2/OpenRCT2/pull/6978/files#diff-79543dc88b0451e0e6b175259ba31effR39

Michał Janiszewski @janisozaur
that's not something we have a use for
right now there can only be one context

Aaron van Geffen @AaronVanGeffen
Might in the future, though, when the ball starts rolling with the object server/provider.

Michał Janiszewski @janisozaur
I'm not sure about that
the object seems reusable and we can just schedule multiple requests that will be handled by it

Aaron van Geffen @AaronVanGeffen
Sure, but you might want to handle them in parallel?

Michał Janiszewski @janisozaur
They may still end up being fine
though requests' code may require checking
the functions seem to be re-entrant, so probably no big deal there

Aaron van Geffen @AaronVanGeffen
Alright

Michał Janiszewski @janisozaur
the only thing we really need to look for is curl_global_init

matrixbot @matrixbot
noname120 right

Michał Janiszewski @janisozaur
I don't like code like this though: https://github.com/OpenRCT2/OpenRCT2/pull/6978/files#diff-1a8ed57416b3c9a1fa4a2c55f8141d43R62
you could end up passing a temporary down that function and I don't know what the json functions do

matrixbot @matrixbot
noname120 true, I don't know about the lib you use but I encountered such issue while using jsmn
noname120 the returned object has pointers to the string in jsmn

matrixbot @matrixbot
noname120 Note that while the jansson's reference of json_loads doesn't say anything about it, the tutorial says the passed string can be freed: https://jansson.readthedocs.io/en/latest/tutorial.html
noname120 > We don’t need the JSON text anymore, so we can free the text variable right after decoding it.
```

**Comment by janisozaur** [comment]
With exception of few minor things, this looks good to me.

It would be nice if you could trace those `json_loads` calls and make sure they get freed.

File: src/openrct2/network/Http.cpp (hunk 9 above)
Comment (IntelOrca) [reply src/openrct2/network/Http.cpp]: We are all C++ now so the method should be changed to return an `std::vector`.

File: src/openrct2/network/Http.h (hunk 10 above)
Comment (tobiaskohlbau) [reply src/openrct2/network/Http.h]: Right it's a relict...

File: src/openrct2/network/Http.cpp (hunk 9 above)
Comment (tobiaskohlbau) [reply src/openrct2/network/Http.cpp]: Didn't want to touch this but I'm fine which changing the function signature to be more C++ like.

File: src/openrct2/Context.cpp (hunk 5 above)
Comment (AaronVanGeffen) [reply src/openrct2/Context.cpp]: This is minor, but while you've reintroduced the aforementioned space, you've also introduced spaces around the hyphen. Please remove those — we'd like to keep the copyright line uniform across files.

Comment (tobiaskohlbau) [reply src/openrct2/Context.cpp]: Dumb me of course I would have to remove that issue from clang-format as well. Will fix it with the next revision.

File: src/openrct2/network/Http.cpp (hunk 9 above)
Comment (tobiaskohlbau) [reply src/openrct2/network/Http.cpp]: I've looked closer at this and I'm not sure why a explicit `download_park` function is required. It's a normal download which as well passed the URL. I've removed it and put the few lines into context.

Comment (janisozaur) [reply src/openrct2/network/Http.cpp]: Wasn't it related somehow to solving the problem of having to download a park first to detect whether it really is a park (sv6)? If you search git history for `tmpnam` maybe you will something useful around there.

**Comment by janisozaur** [comment]
@tobiaskohlbau are you still working on that?

**Comment by tobiaskohlbau** [comment]
I’m aware that there are 2 open PRs by me. I’m busy until the end of February, but in early March I will continue my work on OpenRCT2.

**Comment by AaronVanGeffen** [comment]
I've taken the liberty of rebasing this. The code looks good to me, so I think all that's left to do is some thorough testing.

File: src/openrct2/network/Twitch.cpp [hunk 11]
```diff
@@ -1,4 +1,4 @@
-#pragma region Copyright (c) 2014-2017 OpenRCT2 Developers
+#pragma region Copyright(c) 2014 - 2018 OpenRCT2 Developers
```
Comment (Gymnasiast) [inline src/openrct2/network/Twitch.cpp]: Unnecessary whitespace changes.

File: src/openrct2/core/Json.cpp [hunk 12]
```diff
@@ -1,4 +1,4 @@
-#pragma region Copyright (c) 2014-2017 OpenRCT2 Developers
+#pragma region Copyright(c) 2014 - 2018 OpenRCT2 Developers
```
Comment (Broxzier) [inline src/openrct2/core/Json.cpp]: The spacing here has changed

**Comment by AaronVanGeffen** [comment]
> I find the mixed use of casing strange. The file's name uses 'Http', the namespace is called http, and the scruct uses HTTP. Wouldn't it make more sense to name the namespace network instead?

I agree, that is something to be improved. The header file is also called `http.h` rather than `Http.h`.

Reworking the namespace to `Network` fits the folder name. Effectively, we'd end up with `Network::HttpRequest2`, which sounds good to me.

**Comment by janisozaur** [comment]
I did a quick test on Linux and it seems to work OK.
